R/CallbackTuning.R

```diff
@@ -0,0 +1,135 @@
+    #'   Called in `ObjectiveTuning$eval_many()`.
+    on_eval_after_design = NULL,
+
+    #' @field on_eval_after_benchmark (`function()`)\cr
+    #'   Stage called after hyperparameter configurations are evaluated.
+    #'   Called in `ObjectiveTuning$eval_many()`.
+    on_eval_after_benchmark = NULL,
+
+    #' @field on_eval_before_archive (`function()`)\cr
+    #'   Stage called before performance values are written to the archive.
+    #'   Called in `ObjectiveTuning$eval_many()`.
+    on_eval_before_archive = NULL
+  )
+)
+
+#' @title Create Tuning Callback
+#'
+#' @description
+#' Function to create a [CallbackTuning].
+#' Predefined callbacks are stored in the [dictionary][mlr3misc::Dictionary] [mlr_callbacks] and can be retrieved with [clbk()].
+#'
+#' Tuning callbacks can be called from different stages of tuning process.
+#' The stages are prefixed with `on_*`.
+#'
+#' ```
+#' Start Tuning
+#'      - on_optimization_begin
+#'     Start Tuner Batch
+#'          - on_optimizer_before_eval
+#'         Start Evaluation
+#'              - on_eval_after_design
+#'              - on_eval_after_benchmark
+#'              - on_eval_before_archive
+#'         End Evaluation
+#'          - on_optimizer_after_eval
+#'     End Tuner Batch
+#'      - on_result
+#'      - on_optimization_end
+#' End Tuning
+#' ```
+#'
+#' See also the section on parameters for more information on the stages.
+#' A tuning callback works with [bbotk::ContextOptimization] and [ContextEval].
+#'
+#' @details
+#' A callback can write data to its state (`$state`), e.g. settings that affect the callback itself.
+#' Avoid writing large data the state.
+#' This can slow down the tuning process when the evaluation of configurations is parallelized.
+#'
+#' Tuning callbacks access two different contexts depending on the stage.
+#' The stages `on_eval_after_design`, `on_eval_after_benchmark`, `on_eval_before_archive` access [ContextEval].
+#' This context can be used to customize the evaluation of a batch of hyperparameter configurations.
+#' Changes to the state of callback are lost after the evaluation of a batch and changes to the tuning instance or the tuner are not possible.
+#' Persistent data should be written to the archive via `$aggregated_performance` (see [ContextEval]).
+#' The other stages access [ContextOptimization].
+#' This context can be used to modify the tuning instance, archive, tuner and final result.
+#'
+#' @param id (`character(1)`)\cr
+#'   Identifier for the new instance.
+#' @param label (`character(1)`)\cr
+#'   Label for the new instance.
+#' @param man (`character(1)`)\cr
+#'   String in the format `[pkg]::[topic]` pointing to a manual page for this object.
+#'   The referenced help package can be opened via method `$help()`.
+#' @param on_optimization_begin (`function()`)\cr
+#'   Stage called at the beginning of the optimization.
+#'   Called in `Optimizer$optimize()`.
+#'   The function must have two arguments named `callback` and `context`.
```

Comment:
I think documenting for every argument that it must have two arguments named `callback` and `context` might not be necessary.